lib/Models/SearchProviders/CatalogIndex.ts

```diff
@@ -0,0 +1,131 @@
+import { ObservableMap, action, observable } from "mobx";
+import CatalogIndexReference from "../Catalog/CatalogReferences/CatalogIndexReference";
+import loadJson from "../../Core/loadJson";
+import updateModelFromJson from "../Definition/updateModelFromJson";
+import CommonStrata from "../Definition/CommonStrata";
+import Terria from "../Terria";
+import CatalogIndexReferenceTraits from "../../Traits/TraitsClasses/CatalogIndexReferenceTraits";
+import SearchResult from "./SearchResult";
+
+const { Document } = require("flexsearch");
+
+export interface CatalogIndexFile {
+  [id: string]: CatalogIndexReferenceTraits;
+}
+
+export interface ModelIndex {
+  name: string;
+  knownContainerUniqueIds: string[];
+}
+
+export default class CatalogIndex {
+  private _models: ObservableMap<string, CatalogIndexReference> | undefined;
+
+  get models() {
+    return this._models;
+  }
+  private _searchIndex: any; // Flex-search document index
+
+  get searchIndex() {
+    return this._searchIndex;
+  }
+
+  readonly loadPromise: Promise<void>;
+
+  constructor(private readonly terria: Terria, private readonly url: string) {
+    this.loadPromise = this.loadCatalogIndex();
+  }
+
+  @action
+  private async loadCatalogIndex() {
+    // Load catalog index
+    try {
+      const index = (await loadJson(this.url)) as CatalogIndexFile;
```

Comment:
I don't have a preference here but one thing to consider is whether to load this eagerly or only when the user opens search for the first time and maybe document the choice here?

Reply:
Added a comment